pkg/pubsub/pubsub_init.go

```diff
@@ -0,0 +1,95 @@
+package pubsub
+
+import (
+	"context"
+	"fmt"
+
+	psClient "github.com/open-policy-agent/gatekeeper/pkg/pubsub/client"
+	"github.com/open-policy-agent/gatekeeper/pkg/util"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kubeClient "sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+)
+
+var log = logf.Log.WithName("pubsub")
+
+var _ manager.Runnable = &runner{}
+
+type runner struct {
+	mgr manager.Manager
+}
+
+func AddToManager(m manager.Manager) error {
+	mr := new(m)
+	return m.Add(mr)
+}
+
+func new(mgr manager.Manager) *runner {
+	mr := &runner{
+		mgr: mgr,
+	}
+	return mr
+}
+
+var initiatedTools []psClient.PubSub
+
+// Start implements the Runnable interface.
+func (r *runner) Start(ctx context.Context) error {
+	tools := psClient.Tools()
+	if len(tools) == 0 {
+		log.Info("No pub sub tool is enabled")
+		return nil
+	}
+	log.Info("Initializing pub subs")
+	client := r.mgr.GetClient()
+	for name, newClient := range tools {
+		config := corev1.ConfigMap{ObjectMeta: metav1.ObjectMeta{Name: name, Namespace: util.GetNamespace()}}
+		err := client.Get(ctx, kubeClient.ObjectKey{Name: name, Namespace: util.GetNamespace()}, &config)
+		if err != nil && !errors.IsNotFound(err) {
+			return err
+		}
+		tmp, err := newClient(ctx, config.Data["config"])
+		if err != nil {
+			return err
+		}
+
+		tc, ok := tmp.(psClient.PubSub)
+		if ok {
+			initiatedTools = append(initiatedTools, tc)
+		} else {
+			log.Error(fmt.Errorf("Failed to convert interface"), "Unable to append client")
+		}
+	}
+
+	log.Info("Pub sub clients are initialized without error")
+	return nil
+}
+
+// Publish messages to appropriate endpoints using appropriate configure pubsub tool
+// input: interface to be published, topic/channel name to publish the message in, source/origin of the message (i.e Audit, Validation, etc)
+func Publish(data interface{}, topic string) {
```

Comment:
Having a global function may make testing harder. It could also complicate certain use cases, such as if users wanted to have multiple client sets (honestly most likely for testing, but you never know).

Dependency injection, like we use for things like the constraint framework or the mutation system, would be a more flexible model.

Reply:
I am not following could you share an example?

Reply:
Here is a line from the test for mutation system: https://github.com/open-policy-agent/gatekeeper/blob/476269f47408e4589915f7e6b08e6998a3de38dc/pkg/mutation/system_test.go#L205

Notice that for every test case, we create a new system that starts with a clean slate. This means we do not need to worry about cleanup or stale data complicating the test.

Some ways that may be useful for testing a pubsub interface:

* You could create a client that is guaranteed to fail to test edge cases and not have to add/remove that client
* You could run tests in parallel and not have to worry about them interfering with each other due to shared state

For code:

* It's more clear who has access to the pubsub interface
* Different callers can use different instances of the pubsub interface with different clients (example: say you want to re-use this code for an internal project that does scanning of violations for multiple users, each user gets their own copy of pubsub, which reduces the likelihood that you'll send a violation to the wrong user's queue)

Reply:
Thanks for this! it took me a while to understand how mutation and expansion work but I think now I understand. I made changes that improves the overall approach and follows same pattern as how mutation works